### PR TITLE
Frame: Give Referendum `SubmitOrigin` argument

### PR DESCRIPTION
**Pull request by gavofyork**
When determining whether a submission origin should be allowed, it is useful to be able to inspect the track they are submitting for, since it may be desired to have less-privileged origins be valid for some tracks but not others.

This alters the type bound accordingly. To minimise breakage across the ecosystem, all type bounders in Substrate are updated to support the `EnsureOriginWithArg` trait as well as the `EnsureOrigin` trait. To facilitate this support both in Substrate and downstream repos, a macro allowing `EnsureOrigin` implementors to naively support `EnsureOriginWithArg` is introduced.

## 🚨 Code Changes

If you are using Referenda pallet and find a build error around the `SubmitOrigin` (which is probably a custom origin), then simply wrap the type with `frame_support::AsEnsureOriginWithArg` to get an `EnsureOriginWithArg` impl. E.g.

```rust
impl pallet_referenda::Config for Runtime {
    type SubmitOrigin = MyEnsureOriginImpl;
    // snip
}
```

Would become:

```rust
impl pallet_referenda::Config for Runtime {
    type SubmitOrigin = frame_support::AsEnsureOriginWithArg<MyEnsureOriginImpl>;
    // snip
}
```

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2959545

File: frame/support/src/traits/dispatch.rs [hunk 1]
```diff
@@ -250,6 +318,24 @@ impl<OuterOrigin, L: EnsureOrigin<OuterOrigin>, R: EnsureOrigin<OuterOrigin>>
 		L::try_successful_origin().or_else(|()| R::try_successful_origin())
 	}
 }
+impl<
+		OuterOrigin,
+		L: EnsureOriginWithArg<OuterOrigin, Argument>,
+		R: EnsureOriginWithArg<OuterOrigin, Argument, Success = L::Success>,
```
Comment (ggwpez) [inline frame/support/src/traits/dispatch.rs:324]:
```suggestion
		R: EnsureOriginWithArg<OuterOrigin, Argument>,
```
Otherwise it is not diverse.

**Review by ggwpez — APPROVED** [review]
Just 1 comment and 1 nit.

**Review by sam0x17 — APPROVED** [review]
looks good, would prefer that we make that macro name more clear

**Review by muharem — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2968318

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2968319

**Pull request merged**